pulse counter driver: initial SAUL-compatible implementation

Hyungsin opened the pull request:
This is the initial attempt to provide a SAUL-compatible push-button driver.

Whenever SAUL calls the read function, this driver returns the number of button pushes and initializes the value to zero.

gebart commented:
I like the idea, but I think this is more general than just push buttons. What do you think about naming this a pulse counter?

Hyungsin commented:
@gebart, I can. But then what would be SAUL type? Its type is now SAUL_SENSE_BTN.

gebart commented:
I don't know. I mean what you are reporting is not the actual state of a button, but the number of times it has been pressed. Semantically it seems incorrect to use SENSE_BTN for a counter.

haukepetersen commented:
It would be interesting to get a little more background on the potential usage of this module.

From a code/structure perspective, I see a couple of issues with this PR:
- the code is almost identical to the saul gpio driver
- the driver does allow only for one instance to be active at a time (`push_button_events` is shared)
- the name is misleading, I agree with @gebart that something like pulse counter etc would be more fitting
- the driver does not allow for re-setting the value (which I would think is quite useful?) -> could be mapped to the SAUL write function
- some functions can well be simplified (e.g. `int push_button_read(int16_t *button_events)` does never return an error, so why not make it `int16_t push_button_read(void)`?

gebart commented:
@haukepetersen I agree with your points, but I think it would be even more useful to have the counter atomically reset when read, so that no pulse is ever lost between reading and resetting.
What I mean is that the read call will use something like:
```c
    static const uint32_t zero = 0;
    uint32_t count = 0;
    /* Use atomic operations to avoid messing with IRQ flags */
    __atomic_exchange(the_counter_variable, &zero, &count, __ATOMIC_SEQ_CST);
    return count;
```

and the IRQ callback: 
```c
static void count_up_cb(void* ptr)
{
    /* Use atomic operations to avoid messing with IRQ flags */
    __atomic_fetch_add((uint32_t *)ptr, 1, __ATOMIC_SEQ_CST);
}
```

Hyungsin commented:
@haukepetersen, @gebart, I changed the name to pulse counter and simplified push_button_read function.

Supporting multiple instances would be good. But then, what is your recommendation for implementing that feature? One way could be putting the pulse_count value in each "dev", which makes it hard to reset the value in the reading function since the input "dev" is const.

Hyungsin commented:
@haukepetersen , @gebart, I updated again. 
Now this PR can support multiple pulse counters! :)

cladmi commented:
I have remarks regarding the hardwritten `GPIO_FALLING`.

First, some boards have inverted buttons, meaning pushing the button makes a `GPIO_RISING`. For example on `samr21-xpro`: https://github.com/RIOT-OS/RIOT/blob/2017.10/boards/samr21-xpro/include/gpio_params.h#L47
So it would not have the same behavior on all boards right now.

Also, there may be reason for someone to want to count button release, or both.

What about adding this `gpio_flank_t` as a configuration to `pulse_counter_params_t` ?
Maybe directly, or with a `.flags` mapping if it requires other options later. (Simple or forward compatible).

cladmi commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,67 @@
+ */
+#define PULSE_COUNTER_NUM    (sizeof(pulse_counter_params)/sizeof(pulse_counter_params[0]))
+
+uint16_t pulse_count[PULSE_COUNTER_NUM];
```
I do not really like having this extra `pulse_count` structure here instead of in `auto_init_pulse_counter.c`

Having an indirection is still necessary tough as `saul read` and `saul write` use a `const pulse_counter_t*` pointer, but it could be a pointer and not a table index.

gebart replied:
Good idea.
Also, why is the pulse count variable defined as a 16 bit uint? It does limit the use cases a bit by having the counter roll over at 65535 counts.
I think either 32 bit or using `unsigned` as the type makes more sense here.

Hyungsin replied:
@gebart, does SAUL provide 32-bit output? I set this to 16 bits since I thought SAUL provides this type.

gebart replied:
ah, you are correct. The phydat_t type uses 16 bit ints

Hyungsin replied:
ok :)

Hyungsin replied:
I will take the idea of using pointers. Thanks!

cladmi commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,67 @@
+uint16_t pulse_count[PULSE_COUNTER_NUM];
+
+/* Accumulate pulse count */
+void pulse_counter_trigger(void* arg) {
```
This function should be defined `static`.

Hyungsin replied:
fixed. Thanks!

cladmi commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @return                   0 on success
+ * @return                  -1 on error
+ * @return                  -2 on invalid address
```
The implemented function currently never returns `-2`.

Either the documentation, or the implementation needs to be updated.

Hyungsin replied:
fixed. Thanks!

cladmi commented:
I am not sure if resetting when read is really wanted.
It would break multiple readers using the device at the same time.

Maybe it is "normal" for saul device to be expected to work with only one reader, in that case do not take this into account.

At least, `pulse_counter_read` documentation should state that is resets the counter value.
And maybe even add another function for `read without resetting`.

gebart commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,88 @@
+  * @{
+  */
+typedef struct {
+    uint8_t   id;             /**< pulse counter id to distinguish multiple counters */
```
I suggest putting the count itself here instead of an index to a static array.

gebart replied:
... or as a pointer as suggested by @cladmi

Hyungsin replied:
Yeah, I will use a pointer! Thanks :)

gebart commented on drivers/pulse_counter/include/pulse_counter_params.h:
```diff
@@ -0,0 +1,66 @@
+ * @{
+ */
+#ifndef PULSE_COUNTER_PARAMS_BOARD
+#define PULSE_COUNTER_PARAMS_DEFAULT  { .gpio = GPIO_PIN(PA, 19) }
```
PA is CPU specific, I don't think there is any sensible default possible here (similar to how there is no default gpio_params.h)

Hyungsin replied:
I agree. I will remove DEFAULT parameter so that RIOT incurs compile error without defining BOARD_PARAMS.

gebart commented on drivers/pulse_counter/include/pulse_counter_params.h:
```diff
@@ -0,0 +1,66 @@
+static const saul_reg_info_t pulse_counter_saul_info[] =
+{
+    {
+        .name = "pulse counter",
```
This needs to be configurable as well, otherwise there will be an array out of bounds access for pulse_counter_saul_info[n].name when more than one pulse counter is defined by the config.

gebart replied:
you could merge the name parameter into the pulse_counter_params_t struct, similar to how SAUL GPIO does it.

Hyungsin replied:
Got it. Thanks!

gebart commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,67 @@
+int16_t pulse_counter_read(const pulse_counter_t *dev)
+{
+    int16_t pulse_count_output = pulse_count[dev->id];
+    pulse_counter_reset(dev);
```
This is non-atomic, if a pulse arrives between the line above and this line, the pulse will be lost.

A safer implementation is to use atomic_exchange (see my previous comment for an implementation suggestion).

gebart replied:
...or you could remove the reset when read functionality as suggested by @cladmi. I only wanted to put the idea out there about a reset when read counter.

gebart commented on drivers/pulse_counter/pulse_counter_saul.c:
```diff
@@ -0,0 +1,45 @@
+static int read_pulse_counter(const void *dev, phydat_t *res)
+{
+    res->val[0] = pulse_counter_read((const pulse_counter_t*)dev);
+    memset(&(res->val[1]), 0, 2 * sizeof(int16_t));
```
No need to touch the unused values. The return value 1 tells the caller that the other values are unused and garbage.

Hyungsin replied:
fixed, thanks!

gebart commented on drivers/pulse_counter/pulse_counter_saul.c:
```diff
@@ -0,0 +1,45 @@
+
+static int write_pulse_counter(const void *dev, phydat_t *data)
+{
+    pulse_counter_reset((const pulse_counter_t*)dev);
```
add `(void)data;` since the data parameter is intentionally unused.

Hyungsin replied:
sure, thanks!

gebart commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,88 @@
+ * @brief   Parameters needed for device initialization
+ */
+typedef struct {
+    gpio_t    gpio;            /**< GPIO pin that sensor is connected to */
```
you could merge the name variable into this array to make the _params.h config simpler. (mimic SAUL GPIO)

Hyungsin replied:
fixed, thanks!

Hyungsin commented:
@gebart and @cladmi, I reflected your comments. Please review again.
Only one thing left is resetting thing. I will work on that also.

Hyungsin commented:
@gebart and @cladmi, OK. I made the reading and counting up procedures atomic. Also I added an option to read without reset.

Hyungsin commented:
Any further comment? Is it OK to rebase and squash?

haukepetersen commented on drivers/pulse_counter/include/pulse_counter_params.h:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @brief   PULSE_COUNTER configuration
+ */
+static const pulse_counter_params_t pulse_counter_params[] =
```
Please also add a default configuration (as e.g. discussed in #7519 and applied in e.g. #7955 or #7949) -> this makes it much easier for the CI to build this driver even without board specific configuration.

Hyungsin replied:
sure, thanks :)

haukepetersen commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,75 @@
+/* Initialize pulse counter */
+int pulse_counter_init(pulse_counter_t *dev, const pulse_counter_params_t *params)
+{    
+    dev->p.name = params->name;
```
`memcpy(&dev->p, params, sizeof(pulse_counter_params_t))`?

haukepetersen replied:
or even better: there is no need to keep the params at all, they are not used after `gpio_init_int()` was called. So simply drop the `p` field from the struct I would say.

Hyungsin replied:
sure. Thanks :)

haukepetersen commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,98 @@
+  * @{
+  */
+typedef struct {
+    int16_t   *pulse_count;             /**< pulse counter */
```
I think I am missing something: this field should store that actual counter value right? So why is it a pointer? And where does it point to? Or is the actual pointer used as counter variable? Then the `int16_t` would be strange...

Hyungsin replied:
There are two reasons to make this. 
1. Pointer vs. Value: SAUL read or write function "const" dev as the parameter. So, when we include pulse counter value itself in dev, we cannot change the value while executing the SAUL read function. But using pointer is fine. The pointer points the counter value defined in auto_init_pulse_counter.c.
2. I used int16_t since SAUL's output goes to "phydat_t" which uses int16_t type.

Hyungsin replied:
There are two reasons to make this.

Pointer vs. Value: SAUL read or write function has "const" dev as the parameter. So, when we include pulse counter value itself in dev, we cannot change the value while executing the SAUL read function. But using pointer is fine. The pointer points the counter value defined in auto_init_pulse_counter.c.
I used int16_t since SAUL's output goes to "phydat_t" which uses int16_t type.

haukepetersen replied:
I see, though I don't like it. Seems to me its quite some effort and overhead. As the device descriptor will always be in RAM, I strongly propose to change the drivers interface to use the device descriptor as non-const, and simply use an explicit cast in the SAUL read/write functions
```c
static int read_pulse_counter(const void *dev, phydat_t *res)
{
    res->val[0] = pulse_counter_read_with_reset((pulse_counter_t *)dev);
    ...
```

cladmi replied:
If we go this way, rather than having lying function API, would not it be be better to cast the function pointer when setting the driver ?
 
```
const saul_driver_t pulse_counter_saul_driver = {
    /* read and write functions modify `dev` pointer, but we know it is in ram, so its ok */
    .read  = (saul_read_t)read_pulse_counter,
    .write = (saul_write_t)write_pulse_counter,
    .type  = SAUL_SENSE_COUNT,
};
```

How I see this case, the problem is not that `dev` is modified, as it is in ram, but that `saul_driver_t` takes read/write functions that do not modify dev.
So the API conflict should be solved when setting `saul_driver_t`.


General question, why should saul read and write functions be expected to not modify the pointer ?
It would be more in phase with https://github.com/RIOT-OS/RIOT/pull/7723/files to have read and write taking a non const pointer.

Hyungsin replied:
@haukepetersen and @cladmi, good point. I changed pulse count from a pointer to a value.
I recasted "dev" from void to pulse_counter_t in the device driver, not in the SAUL interface.

haukepetersen commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,75 @@
+/* Return the accumulated pulse counts */
+int16_t pulse_counter_read_without_reset(const pulse_counter_t *dev)
+{
+    return *(dev->pulse_count);
```
you get the value from the memory location the pointer points to, but where does it point? Where is the actual counter variable allocated?

Hyungsin replied:
Please see auto_init_pulse_counter.c. :)

haukepetersen commented on sys/auto_init/saul/auto_init_pulse_counter.c:
```diff
@@ -0,0 +1,77 @@
+    for (unsigned i = 0; i < PULSE_COUNTER_NUM; i++) {
+        LOG_DEBUG("[auto_init_saul] initializing pulse_counter #%u\n", i);
+
+        pulse_counter_devs[i].pulse_count = &pulse_count[i];
```
as stated above: why not save the counter value directly in the device descriptor?

Hyungsin replied:
The answer is above :)

Hyungsin commented:
I reflected @haukepetersen's comments. Please check.

Hyungsin commented:
I updated again, changing pulse_count from a pointer to a value. Thank @haukepetersen and @cladmi!

Hyungsin commented:
Any additional feedback here?

Hyungsin commented:
Squashed and Rebased :)

Hyungsin commented:
@gebart and @cladmi, Please check this and ACK :)

aabadie commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,93 @@
+ * @author      Hyung-Sin <hs.kim@cs.berkeley.edu>
+ */
+
+#ifndef PULSE_COUNTER_H_
```
Please remove the trailing underscore, that's why the CI on Travis is broken (same at the end of this file)

Hyungsin replied:
Thanks, I addressed it! :)

aabadie commented on drivers/Makefile.include:
```diff
@@ -58,6 +58,9 @@ endif
   USEMODULE_INCLUDES += $(RIOTBASE)/drivers/my9221/include
 endif
+ifneq (,$(filter pulse_counter,$(USEMODULE)))
+    USEMODULE_INCLUDES += $(RIOTBASE)/drivers/pulse_counter/include
```
Indent should be 2 spaces

aabadie commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,92 @@
+
+/**
+  * @brief   Device descriptor for a PULSE_COUNTER device
+  * @{
```
With `@brief` you don't need the doxygen block markers `@{` and `@}` I think.

aabadie commented on sys/auto_init/saul/auto_init_pulse_counter.c:
```diff
@@ -0,0 +1,71 @@
+ * @brief   Memory for the SAUL registry entries
+ */
+static saul_reg_t saul_entries[PULSE_COUNTER_NUM];
+
```
Maybe you could add the following here:
```c
/**
 * @brief   Define the number of saul info
 */
#define PULSE_COUNTER_INFO_NUM    (sizeof(pulse_counter_saul_info)/sizeof(pulse_counter_saul_info[0]))
```

Hyungsin replied:
addressed

aabadie commented on sys/auto_init/saul/auto_init_pulse_counter.c:
```diff
@@ -0,0 +1,71 @@
+
+
+void auto_init_pulse_counter(void)
+{
```
And add an assert at the beginning of this function:
```c
assert(PULSE_COUNTER_NUM  == PULSE_COUNTER_INFO_NUM)
```

Hyungsin replied:
addressed

Hyungsin commented:
@aabadie, thanks! Addressed!

aabadie commented on sys/auto_init/saul/auto_init_pulse_counter.c:
```diff
@@ -0,0 +1,74 @@
+/**
+ * @brief   Define the number of configured sensors
+ */
+#define PULSE_COUNTER_NUM    (sizeof(pulse_counter_params)/sizeof(pulse_counter_params[0]))
```
put spaces around `/`

aabadie commented on sys/auto_init/saul/auto_init_pulse_counter.c:
```diff
@@ -0,0 +1,74 @@
+/**
+ * @brief   Define the number of configured sensors
+ */
+#define PULSE_COUNTER_INFO_NUM    (sizeof(pulse_counter_saul_info)/sizeof(pulse_counter_saul_info[0]))
```
put spaces around `/`

aabadie commented on drivers/pulse_counter/include/pulse_counter_params.h:
```diff
@@ -0,0 +1,60 @@
+#endif
+
+#ifndef PULSE_COUNTER_PARAMS
+#define PULSE_COUNTER_PARAMS       { .gpio = GPIO_PIN(PA,18), \
```
`GPIO_PIN(PA,18)` could defined in a seperate default macro before:
```c
#ifndef PULSE_COUNTER_GPIO
#define PULSE_COUNTER_GPIO          GPIO_PIN(PA,18)
#endif
```
same for `.gpio_flank`. 

and then here:
```c
#define PULSE_COUNTER_PARAMS       { .gpio = PULSE_COUNTER_GPIO, \
...
```
The idea is that if a board has a pulse counter then one has just to define `PULSE_COUNTER_GPIO` in `board.h`

aabadie replied:
This comment has not been addressed.

aabadie commented:
And maybe split the commit history in 2:
* drivers/pulse_counter: initial implementation
* sys/auto_init: add SAUL auto init for pulse counter

And another thing (which is blocking for me): add a test application in `tests/`. Otherwise the CI won't build the driver by default.

Hyungsin commented:
@aabadie, done. Thanks!

aabadie commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,69 @@
+#include "debug.h"
+
+/* Accumulate pulse count */
+static void pulse_counter_trigger(void* arg) {
```
The opening curly brace should be on the next line

aabadie commented on tests/driver_pulse_counter/main.c:
```diff
@@ -0,0 +1,60 @@
+#include "saul_reg.h"
+#include "phydat.h"
+
+#define SLEEP       (1000 * 1000U)
```
Use predefined macro `US_PER_S` (or something like that)

aabadie commented on tests/driver_pulse_counter/main.c:
```diff
@@ -0,0 +1,60 @@
+
+#define SLEEP       (1000 * 1000U)
+
+saul_reg_t *sensor_pulse_counter_t;
```
No need to use saul in the test application but rather the public API of the driver.

SAUL can be used when building the `test/saul` or `examples/saul` applications. But I admit the saul adaption won't be built by default by the CI. See #6405

Hyungsin commented:
@aabadie, addressed your comments!

aabadie commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,90 @@
+ */
+
+/**
+ * @defgroup    drivers_pulse_counter sensor
```
extra `sensor`

Hyungsin replied:
addressed

aabadie commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,90 @@
+ * @brief   Initialize an PULSE_COUNTER device
+ *
+ * @param[out] dev          device descriptor
+ * @param[in] params        GPIO bus the device is connected to
```
doxygen comment and parameter mismatch

Hyungsin replied:
addressed

aabadie commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,90 @@
+ *
+ * @param[in]  dev          device descriptor of sensor
+ *
+ * @return                  Accumunlated pulse counts
```
Typo: Accumu**n**lated

Hyungsin replied:
addressed

aabadie commented on drivers/include/pulse_counter.h:
```diff
@@ -0,0 +1,90 @@
+ *
+ * @param[in]  dev          device descriptor of sensor
+ *
+ * @return                  Accumunlated pulse counts
```
Typo

Hyungsin replied:
addressed

aabadie commented on drivers/pulse_counter/include/pulse_counter_params.h:
```diff
@@ -0,0 +1,66 @@
+#define PULSE_COUNTER_GPIO_FLANK GPIO_FALLING
+#endif
+
+#define PULSE_COUNTER_PARAMS       { .gpio = PULSE_COUNTER_GPIO, \
```
You can wrap this with an `#ifndef PULSE_COUNTER_PARAMS .. #endif`

Hyungsin replied:
addressed

aabadie commented on drivers/pulse_counter/pulse_counter_saul.c:
```diff
@@ -0,0 +1,45 @@
+ */
+
+/**
+ * @ingroup     driver_pulse_counter
```
typo in group name: should be driver**s**_pulse_counter

Hyungsin replied:
addressed

aabadie commented on tests/driver_pulse_counter/main.c:
```diff
@@ -0,0 +1,53 @@
+int main(void)
+{
+    pulse_counter_t dev;
+    int16_t count;
```
why a signed value ? I thought a count value is supposed to be positive.

Hyungsin replied:
pulse counter driver returns pulse count value as the int16_t type for SAUL compatibility.

aabadie commented on tests/driver_pulse_counter/Makefile:
```diff
@@ -0,0 +1,8 @@
+include ../Makefile.tests_common
+
+FEATURES_REQUIRED = periph_gpio
```
I don't think this will be required if you add this as a depenency of the pulse_counter module in `drivers/Makefile.dep`

Hyungsin replied:
addressed

Hyungsin commented:
@aabadie, updated again!

Hyungsin commented:
Can we merge this for this release if there are no additional comments?

aabadie commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,70 @@
+/* Accumulate pulse count */
+static void pulse_counter_trigger(void* arg)
+{
+    pulse_counter_t *dev = (pulse_counter_t*)arg;
```
missing space before `*`

Hyungsin replied:
addressed

aabadie replied:
Thanks, please squash directly

aabadie commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,70 @@
+#include "debug.h"
+
+/* Accumulate pulse count */
+static void pulse_counter_trigger(void* arg)
```
parameter should be formatted like this: `void *arg`

Hyungsin replied:
addressed

Hyungsin commented:
@aabadie, Done! :)

PeterKietzmann commented:
@gebart @cladmi would you remove your change requests now?!

gebart commented:
I think my comments have been addressed

aabadie commented:
ping @cladmi, just approve (if it's the case of course)

Hyungsin commented:
@gebart, Murdock is not happy with msp430-based boards. Do you know the reason?

Hyungsin commented:
It seems like msp430-based boards do not support atomic operations well, resulting in errors while compiling test application. I think this issue is independent from this PR.
Anyone can work on the msp430 driver side? or I can simply remove the test application from this PR to make Murdock happy.

gebart commented:
The msp430 compiler is the ancient mspgcc 4.7 which does not support modern atomic operations, but only the older __sync builtins. I suggest that the msp430 is blacklisted in the test until someone updates the msp430 support to the modern msp430-elf-gcc toolchain.

Hyungsin commented:
@gebart, @aabadie. Thank you! Now Murdock is happy.
@cladmi, please approve this PR when you have time and if it is the case :)

Hyungsin commented:
@cladmi, ping?

gebart commented on drivers/pulse_counter/pulse_counter.c:
```diff
@@ -0,0 +1,70 @@
+    pulse_counter_t *dev = (pulse_counter_t *)arg;
+
+    /* Use atomic operations to avoid messing with IRQ flags */
+    __atomic_fetch_add(&(dev->pulse_count), 1, __ATOMIC_SEQ_CST);
```
It should be possible to remove the double underscores prefix if RIOT moves to the C11 standard (is currently C99).
